distributed/joblib.py

```diff
@@ -136,11 +163,31 @@ def callback_wrapper():
         return future
 
     def abort_everything(self, ensure_ready=True):
-        # Tell the client to cancel any task submitted via this instance
-        # as joblib.Parallel will never access those results.
+        """ Tell the client to cancel any task submitted via this instance
+
+        joblib.Parallel will never access those results
+        """
         self.client.cancel(self.futures)
         self.futures.clear()
 
+    @contextlib.contextmanager
+    def retrieval_context(self):
+        """Override ParallelBackendBase.retrieval_context to avoid deadlocks.
+
+        This removes thread from the worker's thread pool (using 'secede').
+        Seceding avoids deadlock in nested parallelism settings.
+        """
+        # See 'joblib.Parallel.__call__' and 'joblib.Parallel.retrieve' for how
+        # this is used.
+        try:
+            secede()
+        except ValueError:
+            # We are not a worker, i.e. not in nested parallelism.
+            pass
+
+        yield
+        # No need to rejoin here
```

Comment:
Why? Won't that create too many threads on the distributed worker nodes?

Also, is it valid to call `secede()` several times in the same thread? That would happen if there are several consecutive calls to `Parallel()` in a nested function. We should probably add a test for that case.

Reply:
If we're likely to do more work in this thread after completing then yes, we should `rejoin`.

@ogrisel I suspect that you're thinking of cases like the following?

```python
results = Parallel()(...)

# do work

results2 = Parallel()(...)
```

Reply:
yes